Adopt more smart pointers in WebCore/accessibility

annevk opened the pull request:
#### bac3e33a72951eeb863da9d39142b436bee6a232
<pre>
Adopt more smart pointers in WebCore/accessibility
<a href="https://bugs.webkit.org/show_bug.cgi?id=294496">https://bugs.webkit.org/show_bug.cgi?id=294496</a>

Reviewed by Tyler Wilcock.

As suggested by the static analyzer and

<a href="https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines">https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines</a>
Canonical link: <a href="https://commits.webkit.org/296263@main">https://commits.webkit.org/296263@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e3e317026d77187b0de1b37e0932fce8e69b721f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108016 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27681 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18100 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113226 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58533 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/109979 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28376 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36230 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/81998 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/110964 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22490 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97316 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62430 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21902 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15452 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/57973 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91843 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15515 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116353 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35087 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25831 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91033 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35463 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93594 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90827 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23138 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35726 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13481 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/30897 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/34986 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40540 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34729 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38087 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36390 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/ca10a6ae522137320b6c22b6bfe5302dd284a205. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46753

webkit-early-warning-system commented:
EWS run on current version of this PR (hash https://github.com/WebKit/WebKit/commit/ca10a6ae522137320b6c22b6bfe5302dd284a205)<details>

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/107865 "Build is in progress. Recent messages:OS: Sequoia (15.5), Xcode: 16.4") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27532 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/17951 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113079 "Build is in progress. Recent messages:") | [⏳ 🛠 win ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/109831 "Build is in progress. Recent messages:") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28228 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36076 "Build is in progress. Recent messages:OS: Sequoia (15.5), Xcode: 16.4") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/5/builds/113079 "Build is in progress. Recent messages:") | [⏳ 🧪 win-tests ](https://ews-build.webkit.org/#/builders/Win-Build-EWS "Waiting in queue, processing has not started yet") 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/110810 "Build is in progress. Recent messages:OS: Sequoia (15.5), Xcode: 16.4") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/130/builds/28228 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-mac](https://ews-build.webkit.org/#/builders/138/builds/17951 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-wpe](https://ews-build.webkit.org/#/builders/5/builds/113079 "Build is in progress. Recent messages:") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/130/builds/28228 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/138/builds/17951 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🛠 🧪 jsc ](https://ews-build.webkit.org/#/builders/JSC-Tests-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/130/builds/28228 "Build is in progress. Recent messages:Running configuration") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/138/builds/17951 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116200 "Build is in progress. Recent messages:Running configuration") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/34934 "Build is in progress. Recent messages:OS: Sequoia (15.5), Xcode: 16.4") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/123/builds/36076 "Build is in progress. Recent messages:OS: Sequoia (15.5), Xcode: 16.4") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/2/builds/116200 "Build is in progress. Recent messages:Running configuration") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35310 "Build is in progress. Recent messages:OS: Sequoia (15.5), Xcode: 16.4") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/138/builds/17951 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-gtk](https://ews-build.webkit.org/#/builders/2/builds/116200 "Build is in progress. Recent messages:Running configuration") | 
| | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/138/builds/17951 "Build is in progress. Recent messages:OS: Sonoma (14.7.3), Xcode: 15.4") | [⏳ 🛠 playstation ](https://ews-build.webkit.org/#/builders/PlayStation-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🛠 tv ](https://ews-build.webkit.org/#/builders/tvOS-18-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-safer-cpp ](https://ews-build.webkit.org/#/builders/macOS-Safer-CPP-Checks-EWS "Waiting in queue, processing has not started yet") | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34580 "Build is in progress. Recent messages:Running configuration") | | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/37941 "Build is in progress. Recent messages:OS: Sequoia (15.5), Xcode: 16.4") | | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36242 "Build is in progress. Recent messages:") | | | </details>

twilco commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -634,7 +634,7 @@ AXCoreObject::AccessibilityChildrenVector AXCoreObject::selectedChildren()
         return selectedRows();
     case AccessibilityRole::TabList:
-        if (auto* selectedTab = selectedTabItem())
+        if (RefPtr selectedTab = selectedTabItem())
             return { { *selectedTab } };
```
Just so I understand...why does this need to be a smart pointer? All we do is deref it and put it into a vector. Is it that the static analyzer considers the act of constructing a vector non-trivial, thus being able to theoretically invalidate the raw pointer?

Also, `AccessibilityChildrenVector` is `Vector<Ref<AXCoreObject>>`. Thus, if we do in fact actually need a `RefPtr` here, I think we can save ref-count churn by using `RefPtr::releaseNonNull` here rather than `*selectedTab`.

twilco commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -643,7 +643,7 @@ AXCoreObject::AccessibilityChildrenVector AXCoreObject::selectedChildren()
         if (auto* descendant = activeDescendant())
             return { { *descendant } };
-        if (auto* focusedElement = focusedUIElement())
+        if (RefPtr focusedElement = focusedUIElement())
             return { { *focusedElement } };
```
I think `RefPtr::releaseNonNull` over deref'ing saves ref-count churn here?

twilco commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -1016,7 +1016,7 @@ AXCoreObject::AccessibilityChildrenVector AXCoreObject::columnHeaders()
         auto columns = this->columns();
         for (const auto& column : columns) {
-            if (auto* header = column->columnHeader())
+            if (RefPtr header = column->columnHeader())
                 headers.append(*header);
```
I think `RefPtr::releaseNonNull` over deref'ing saves ref-count churn here?

twilco commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -1540,8 +1540,8 @@ void AXCoreObject::appendRadioButtonGroupMembers(AccessibilityChildrenVector& li
     if (role() == AccessibilityRole::Row) {
-        if (auto* table = exposedTableAncestor())
-            return table;
+        if (RefPtr table = exposedTableAncestor())
+            return table.get();
```
Another place where I don't understand why a smart pointer is necessary. All we do is null-check the raw pointer. If this _is_ actually necessary, I'm fully willing to accept this is my lack of understanding of safety, or alternatively willing to accept that the static analyzer is not very smart...just want to know which one it is 🙂

twilco commented on Source/WebCore/accessibility/AXCoreObject.h:
```diff
@@ -1331,7 +1331,7 @@ class AXCoreObject : public RefCountedAndCanMakeWeakPtr<AXCoreObject> {
     std::optional<AXID> idOfNextSiblingIncludingIgnoredOrParent() const
     {
-        auto* object = nextSiblingIncludingIgnoredOrParent();
+        RefPtr object = nextSiblingIncludingIgnoredOrParent();
         return object ? std::optional(object->objectID()) : std::nullopt;
```
I don't think this one is necessary (I created this function very recently specifically to appease the static analyzer, since only trivial functions are called within)

annevk replied:
https://build.webkit.org/results/Apple-Sequoia-Safer-CPP-Checks/296192%40main%20%283840%29/scan-build-output/StaticAnalyzer/WebCore/StaticAnalyzerReports/report-AXCoreObject.h-idOfNextSiblingIncludingIgnoredOrParent-1-131085.html#EndPath

twilco commented on Source/WebCore/accessibility/AXCoreObject.h:
```diff
@@ -1813,19 +1813,19 @@ T* highestEditableAncestor(T& startObject)
     for (const auto& relatedObject : relatedObjects) {
-        auto* ancestor = findAncestor(descendant, false, [&relatedObject] (const auto& ancestor) {
+        RefPtr ancestor = findAncestor(descendant, false, [&relatedObject](const auto& ancestor) {
             return relatedObject.get() == &ancestor;
         });
         if (ancestor)
-            return ancestor;
+            return ancestor.get();
```
Another questionably necessary one — we're just null-checking it?

twilco commented on Source/WebCore/accessibility/AXCoreObject.h:
```diff
@@ -1853,8 +1853,8 @@ AXCoreObject* findUnignoredDescendant(T& object, bool includeSelf, const F& matc
     for (Ref child : object.childrenIncludingIgnored()) {
-        if (auto* descendant = findUnignoredDescendant(child.get(), /* includeSelf */ true, matches))
-            return descendant;
+        if (RefPtr descendant = findUnignoredDescendant(child.get(), /* includeSelf */ true, matches))
+            return descendant.get();
```
Another questionably necessary one — we're just null-checking it?

twilco commented on Source/WebCore/accessibility/AXLogger.cpp:
```diff
@@ -1303,7 +1303,7 @@ void streamAXCoreObject(TextStream& stream, const AXCoreObject& object, const Op
 
     if (options & AXStreamOptions::ParentID) {
-        auto* parent = object.parentObjectUnignored();
+        RefPtr parent = object.parentObjectUnignored();
         stream.dumpProperty("parentID"_s, parent ? parent->objectID().toUInt64() : 0);
```
This only calls a trivial function, is this necessary?

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -899,8 +899,8 @@ void AXObjectCache::cacheAndInitializeWrapper(AccessibilityObject& newObject, DO
 {
-    if (auto* object = get(widget))
-        return object;
+    if (RefPtr object = get(widget))
+        return object.get();
```
Another questionably necessary one — we're just null-checking it?

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -922,10 +922,10 @@ AccessibilityObject* AXObjectCache::getOrCreate(Widget& widget)
 {
-    if (auto* object = get(node))
-        return object;
+    if (RefPtr object = get(node))
+        return object.get();
```
Another questionably necessary one — we're just null-checking it?

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -975,8 +975,8 @@ AccessibilityObject* AXObjectCache::getOrCreate(Node& node, IsPartOfRelation isP
     // The object may have already been created during relations update.
-    if (auto* object = get(node))
-        return object;
+    if (RefPtr object = get(node))
+        return object.get();
```
Another questionably necessary one — we're just null-checking it?

twilco commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -999,8 +999,8 @@ AccessibilityObject* AXObjectCache::getOrCreate(Node& node, IsPartOfRelation isP
 {
-    if (auto* object = get(renderer))
-        return object;
+    if (RefPtr object = get(renderer))
+        return object.get();
```
Another questionably necessary one — we're just null-checking it?

cdumez commented on Source/WebCore/accessibility/AXCoreObject.cpp:
```diff
@@ -453,8 +453,8 @@ AXCoreObject* AXCoreObject::nextUnignoredSibling(bool updateChildrenIfNeeded, AX
     RefPtr parent = parentObject();
-    if (auto* nextSibling = nextSiblingIncludingIgnored(/* updateChildrenIfNeeded */ true))
-        return nextSibling;
+    if (RefPtr nextSibling = nextSiblingIncludingIgnored(/* updateChildrenIfNeeded */ true))
+        return nextSibling.get();
```
cc @rniwa

annevk replied:
I'm pretty sure all of these are currently flagged by the static analyzer.

https://build.webkit.org/results/Apple-Sequoia-Safer-CPP-Checks/296192%40main%20%283840%29/scan-build-output/StaticAnalyzer/WebCore/index.html is what I used to go through these files. I only selected "Uncounted raw pointer or reference not provably backed by ref-counted variable" to keep the changes somewhat mechanical and easier to review.

annevk replied:
Thanks, I corrected the `releaseNonNull()` and `WTFMove` comments where possible. I double checked a number of the ones you found questionable with the static analyzer results and they were flagged.

It might be that in the future we can undo some of this again. I've been doing a bunch of that recently with protected methods that were introduced for member variables that could have been const instead. That seems okay.

twilco replied:
Thank you for taking care of this 🙏

annevk committed:
Adopt more smart pointers in WebCore/accessibility
https://bugs.webkit.org/show_bug.cgi?id=294496

Reviewed by Tyler Wilcock.

As suggested by the static analyzer and

https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines
Canonical link: https://commits.webkit.org/296263@main

webkit-commit-queue commented:
Committed 296263@main (bac3e33a7295): <https://commits.webkit.org/296263@main>

Reviewed commits have been landed. Closing PR #46753 and removing active labels.
